shared/modules/metametrics.js

```diff
@@ -0,0 +1,199 @@
+ * @param {boolean} defaultMetaMetricsExclusion - Used to set the default state of exludeMetaMetricsId
+ * @returns {(payload: MetaMetricsEventPayload) => Promise<void>} - function to track an event
+ */
+export function getTrackMetaMetricsEvent (
+  metamaskVersion,
+  getDynamicState,
+  defaultMetaMetricsExclusion = false,
+) {
+  const version = process.env.METAMASK_ENVIRONMENT === 'production'
+    ? metamaskVersion
+    : `${metamaskVersion}-${process.env.METAMASK_ENVIRONMENT}`
+
+  return function trackMetaMetricsEvent ({
+    event,
+    category,
+    isOptIn,
+    properties = {},
+    revenue,
+    currency,
+    value,
+    excludeMetaMetricsId: excludeId,
+    additionalContext = {},
+  }) {
+    if (!event || !category) {
+      throw new Error('Must specify event and category.')
+    }
+    const {
+      participateInMetaMetrics,
+      context: providedContext,
+      metaMetricsId,
+      environmentType,
+      chainId,
+      network,
+      metaMetricsSendCount,
+    } = getDynamicState()
+
+    let excludeMetaMetricsId = excludeId ?? defaultMetaMetricsExclusion
+
+    // This is carried over from the old implementation, and will likely need
+    // to be updated to work with the new tracking plan. I think we should use
+    // a config setting for this instead of trying to match the event name
+    const isSendFlow = Boolean(event.match(/^send|^confirm/u))
+    if (isSendFlow && metaMetricsSendCount && !sendCountIsTrackable(metaMetricsSendCount + 1)) {
+      excludeMetaMetricsId = true
+    }
+
+    if (!participateInMetaMetrics && !isOptIn) {
+      return Promise.resolve()
+    }
+
+    const context = merge(providedContext, additionalContext, {
+      app: {
+        name: 'MetaMask Extension',
+        version,
+      },
+      userAgent: window.navigator.userAgent,
+    })
+
+    const trackOptions = {
+      event,
+      properties: {
+        // These values are omitted from properties because they have special meaning
+        // in segment. https://segment.com/docs/connections/spec/track/#properties.
+        // to avoid accidentally using these inappropriately, you must add them as top
+        // level properties on the event payload.
+        ...omit(properties, ['revenue', 'currency', 'value']),
+        revenue,
+        value,
+        currency,
+        category,
+        network,
+        chain_id: chainId,
+        environment_type: environmentType,
+      },
+      context,
+    }
+
+    if (excludeMetaMetricsId) {
+      trackOptions.anonymousId = METAMETRICS_ANONYMOUS_ID
+    } else {
+      trackOptions.userId = metaMetricsId
+    }
+
+    segment.track(trackOptions)
+
+    /**
+     * Some events will want to wait for the track before proceeding. Segment does
+     * things a little differently depending on environment (flushing, vs batching)
+     * so the callback option on .track may behave oddly. It is safe to add a small timeout
```

Comment:
This.... is unfortunate. We have situations where we need to send an event immediately before closing the notification window, which ends the process :grimacing:

Hmm. I guess we can cross that bridge when we come to it, perhaps by sending those from the background.

Reply:
That said, I'm not sure this is particularly useful? I don't think the timing is critical in any situation except those where the process is about to end, and in that situation it sounds like this still doesn't provide any guarantee.

Reply:
that all makes sense to me. One other factor that made me go this route was that we already had places that were using `.then` on the analytics call.

Does the beforeunload event handler work in the background as well? I just saw node-analytics has a manual `.flush` that I can call that'll resolve the segment issue.

Reply:
Interesting! That sounds promising.

I'm not even sure if `beforeunload` works in the popup. I haven't verified it anyway. I'd also be curious to know whether it works in the background.

Oh and this whole thread was non-blocking for this PR - we can improve this later if that's easier.

Reply:
`metametrics-opt-in.component.js` is the example I was thinking of. I wanted to confirm my memory wasn't playing tricks on me.

```js
const promise = participateInMetaMetrics === true
  ? metricsEvent({
    eventOpts: {
      category: 'Onboarding',
      action: 'Metrics Option',
      name: 'Metrics Opt Out',
    },
    isOptIn: true,
  })
  : Promise.resolve()

promise
  .then(() => {
    history.push(nextRoute)
  })
```

Reply:
Interesting - actually now that you mention it, I can see the case for blocking on that in some cases. It is particularly important that we get that particular event.

Reply:
I did some testing and 'beforeunload' does not happen for notification windows. 'unload', does though but we get no guarantees that `segment.flush` will finish and do its thing. I have made a call to make `flushAt: 1` for all environments instead of just development. It doesn't seem like we'd be capturing all of the events in notifications.

Reply:
This is represented in the latest work

Reply:
Good to know! This seems like a great solution.

I can understand why they'd want to batch events together by default but our usage is far from typical. We can turn the batching back on in the background specifically (and maybe in the fullscreen UI?) if this ever becomes a concern.